src/dhtproto/client/helper/ChannelSerializer.d

```diff
@@ -0,0 +1,382 @@
+        this.fork = new SafeFork(&this.forkEntryPoint);
+    }
+
+    /***************************************************************************
+
+        Dumps the provided associative array to disk.
+
+        Params:
+            aa = associative array to dump
+            block = if true, performs a blocking dump; otherwise performs an
+                asynchronous dump (see module header)
+
+        Returns:
+            true if a dump occurred; false if it was skipped
+
+    ***************************************************************************/
+
+    public bool dump ( Contiguous!(S)[hash_t] aa, bool block = false )
+    {
+        void dumpAA ( void delegate ( hash_t, ref Contiguous!(S) ) dump_record )
+        {
+            foreach ( k, v; aa)
+                dump_record(k, v);
+        }
+        return this.dump(&dumpAA, block);
+    }
+
+    /***************************************************************************
+
+        Dumps the provided map to disk.
+
+        Params:
+            map = map to dump
+            block = if true, performs a blocking dump; otherwise performs an
+                asynchronous dump (see module header)
+
+        Returns:
+            true if a dump occurred; false if it was skipped
+
+    ***************************************************************************/
+
+    public bool dump ( Map!(Contiguous!(S), hash_t) map, bool block = false )
+    {
+        void dumpMap ( void delegate ( hash_t, ref Contiguous!(S) ) dump_record )
+        {
+            foreach ( k, v; map)
+                dump_record(k, v);
+        }
+        return this.dump(&dumpMap, block);
+    }
+
+    /***************************************************************************
+
+        Dumps an arbitrary key-value data structure to disk.
+
+        Params:
+            record_dump_dg = delegate to receive a delegate to be called once
+                per record to dump
+            block = if true, performs a blocking dump; otherwise performs an
+                asynchronous dump (see module header)
+
+        Returns:
+            true if a dump occurred; false if it was skipped
+
+    ***************************************************************************/
+
+    public bool dump ( DumpRecordDg record_dump_dg, bool block = false )
+    {
+        this.new_record_dump_dg = record_dump_dg;
+        return this.fork.call(block);
+    }
+
+    // TODO: could add dump methods for other K/V containers that have a means
+    // of iterating over all elements
+
+    /***************************************************************************
+
+        Loads an associative array from disk.
+
+        Params:
+            aa = associative array to load into
+
+    ***************************************************************************/
+
+    public void load ( ref Contiguous!(S)[hash_t] aa )
+    {
+        void loadRecord ( hash_t k, ref Contiguous!(S) v )
+        {
+            aa[k] = v;
```

Comment:
Is there any memory management you have to do here? For example, when loading keys/values which have arrays, would you have to manually duplicated them, or is this already handled somewhere else?

Reply:
In Alligator we used to have to do this manually, because we used the MapSerializer which uses a buffer for the keys / values, and there we had to copy the arrays manually.

Reply:
I believe it's handled by the deserialization code below (https://github.com/sociomantic-tsunami/dhtproto/pull/162/commits/4a31c38d3e3af1f71b07b69741992af9c9394137#diff-5c38ebc93bed0b168f5f68fe43193a84R240), but we should definitely confirm that.

Reply:
Yep. Looks like `v` (https://github.com/sociomantic-tsunami/dhtproto/pull/162/commits/4a31c38d3e3af1f71b07b69741992af9c9394137#diff-5c38ebc93bed0b168f5f68fe43193a84R236) will initially contain an empty buffer, which will be allocated and filled with the deserialized struct. The reference to this is then copied into the AA or map.

Reply:
This is one of the nice things about working with `Contiguous!(S)`, by the way.

Reply:
Nice.